MobileWallet/Backup/ICloudBackup.swift

```diff
@@ -0,0 +1,342 @@
+            self?.query.enableUpdates()
+        })
+    }
+
+    func restoreWallet(completion: () -> Void) throws {
+        Keychain.logout()
+        let wallets = try existsWallets()
+        guard let firstWallet = wallets.first else { return }
+
+        let dbDirectory = TariLib.shared.databaseDirectory
+        try FileManager.default.createDirectory(at: dbDirectory, withIntermediateDirectories: true, attributes: nil)
+
+        try ICloudBackup.shared.restoreBackup(walletFolder: firstWallet, to: dbDirectory) {
+            let privateKey = PrivateKey()
+            let (privateKeyHex, hexError) = privateKey.hex
+
+            if hexError != nil {
+                try FileManager.default.removeItem(at: dbDirectory)
+                throw ICloudBackupWalletError.privateKeyError
+            }
+
+            //Save to keychain and then ensure it's there (If one exists already in the keychain this will overwrite it)
+            TariLib.shared.storedPrivateKey = privateKeyHex
+            guard TariLib.shared.storedPrivateKey != nil else {
+                TariLogger.error("Failed to save private key to keychain")
+                try FileManager.default.removeItem(at: dbDirectory)
+                throw ICloudBackupWalletError.privateKeyError
+            }
+            completion()
+        }
+    }
+}
+
+// MARK: - private methods
+extension ICloudBackup {
+    private func restoreBackup(walletFolder: String, to directory: URL, completion:(() throws -> Void)) throws {
+        try downloadBackup(walletFolder: walletFolder) { zippedBackup in
+            try unzipBackup(url: zippedBackup, to: directory)
+            try completion()
+        }
+    }
+
+    private func downloadBackup(walletFolder: String, completion:((_ url: URL) throws -> Void)) throws {
+        let fileManager = FileManager.default
+        let icloudFolderURL = try iCloudDirectory().appendingPathComponent(walletFolder)
+        let urls = try fileManager.contentsOfDirectory(atURL: icloudFolderURL, sortedBy: .created, options: [])
+
+        if let backupUrl = urls.last {
+            var lastPathComponent = backupUrl.lastPathComponent
+            let folderPath = backupUrl.deletingLastPathComponent().path
+            // if the last path component contains the “.icloud” extension. If yes the file is not on the device else the file is already downloaded.
+
+            if lastPathComponent.contains(".icloud") {
+                lastPathComponent.removeFirst()
+                let downloadedFilePath = folderPath + "/" + lastPathComponent.replacingOccurrences(of: ".icloud", with: "")
+                var isDownloaded = false
+                try fileManager.startDownloadingUbiquitousItem(at: backupUrl)
+
+                while !isDownloaded {
+                    if fileManager.fileExists(atPath: downloadedFilePath) {
+                        isDownloaded = true
+                        try completion(URL(fileURLWithPath: downloadedFilePath))
+                    }
```

Comment:
You can insert a small wait here to avoid the execution of the body of the `while` block hundreds of times non-stop.

Reply:
done